tpls/desul/include/desul/atomics/Compare_Exchange_OpenACC.hpp

```diff
@@ -0,0 +1,266 @@
+  unsigned long long int return_val = atomicExch(reinterpret_cast<unsigned long long int*>(dest),
+                                       reinterpret_cast<unsigned long long int&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 4), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(float) == 4,
+                "this function assumes an float is 32-bit");
+  float return_val = atomicExch(reinterpret_cast<float*>(dest),
+                                       reinterpret_cast<float&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#ifndef DESUL_CUDA_ARCH_IS_PRE_PASCAL
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope) {
+  static_assert(sizeof(double) == 8,
+                "this function assumes an double is 64-bit");
+  double return_val = atomicExch(reinterpret_cast<double*>(dest),
+                                       reinterpret_cast<double&>(value));
+  return reinterpret_cast<T&>(return_val);
+}
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_floating_point<T>::value && (sizeof(T) == 8), T> 
+	device_atomic_exchange(T* const dest, T value, MemoryOrder, MemoryScope scope) {
+  if (acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_exchange(): Not supported atomic "
+                  "operation in the OpenACC backend\n");
+  }
+  // Acquire a lock for the address
+  //while (!lock_address_openacc((void*)dest, scope)) {
+  //}
+  //device_atomic_thread_fence(MemoryOrderAcquire(), scope);
+  T return_val = *dest;
+  *dest = value;
+  //device_atomic_thread_fence(MemoryOrderRelease(), scope);
+  //unlock_address_openacc((void*)dest, scope);
+  return return_val;
+}
+
+#endif
+
+#else
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<std::is_arithmetic<T>::value, T> device_atomic_exchange(T* dest, T value, MemoryOrder, MemoryScope) {
+  T return_val;
+#pragma acc atomic capture
+  {
+    return_val = *dest;
+    *dest = value;
+  }
+  return return_val;
+}
+
+#endif
+
+
+#ifdef KOKKOS_COMPILER_NVHPC
+
+#pragma acc routine seq
+template <class T, class MemoryOrder, class MemoryScope>
+std::enable_if_t<!std::is_arithmetic<T>::value || ((sizeof(T) != 4) && (sizeof(T) != 8)), T> 
+	device_atomic_compare_exchange(T* dest, T compare, T value, MemoryOrder, MemoryScope scope) {
+  T current_val = *dest;
+  if (acc_on_device(acc_device_not_host)) {
+    printf("Kokkos Error in device_atomic_compare_exchange(): Not supported atomic "
```

Comment:
This is `desul` not `Kokkos`.

Reply:
"Kokkos Error" is changed to "DESUL error".